### ISSUE TITLE
Serialization of private structs in a class with version

### ISSUE THREAD

**Issue by cstamatopoulos** [USCiLab/cereal#270]
Consider this small class that I would like to serialize

`
class Foo
{
public:

	void readSettings(const std::string& settings_file)
	{
		try
		{
			std::ifstream is(settings_file);

			if (is)
			{
				cereal::JSONInputArchive archive(is);
				archive(cereal::make_nvp("Bar", bar1));
				archive(cereal::make_nvp("Bar1", bar2));
			}
		}
		catch (const std::exception&)
		{

		}
	}

	void writeSettings(const std::string& settings_file)
	{
		try
		{
			std::ofstream os(settings_file);

			if (os)
			{
				cereal::JSONOutputArchive archive(os);
				archive(cereal::make_nvp("Bar", bar1));
				archive(cereal::make_nvp("Bar1", bar2));
			}
		}
		catch (const std::exception&)
		{

		}
	}

private:
	uint8_t x, y;
	float z;

	struct Bar
	{
		uint8_t x, y;
		float z;

		template<class Archive>
		void serialize(Archive & archive, std::uint32_t const version)
		{
			archive(cereal::make_nvp("x", x),
					cereal::make_nvp("y", y),
					cereal::make_nvp("z", z));

		}
	} bar1;

	struct Bar1
	{
		uint32_t x, y;
		double z;

		template<class Archive>
		void serialize(Archive & archive, std::uint32_t const version)
		{
			archive(cereal::make_nvp("x", x),
					cereal::make_nvp("y", y),
					cereal::make_nvp("z", z));

		}
	} bar2;

	friend class cereal::access;
	friend class cereal::detail::Version<Bar>;
	friend class cereal::detail::Version<Bar1>;
};

CEREAL_CLASS_VERSION(Foo::Bar, 1);
CEREAL_CLASS_VERSION(Foo::Bar1, 1);`

After expanding the macro of CEREAL_CLASS_VERSION I saw that there was a struct Version so the only way for it to see the internal part of Foo was to befriend it. I guess I should be using things that are in detail namespace outside cereal but there is no other way since it would not compile otherwise.

Then I get hundrerds of warnings.
 warning C4099: 'cereal::detail::Version<Foo::Bar,cereal::detail::`anonymous-namespace'::version_binding_tag>': type name first seen using 'struct' now seen using 'class'

That works but it feels cludgy.
I guess I should not be befriending Version but what is the alternative ?

**Comment by AzothAmmo** [comment]
You are right it isn't a very elegant workaround. I'm not sure that we can actually solve this without changing how `CEREAL_CLASS_VERSION` works. Once we're outside of the scope of the class I don't think we can refer to the private member directly without hitting some problems. If we moved the version macro into class scope we could generate friend declarations automatically, but this might have some other problems I haven't thought about.
